### ISSUE TITLE
RIGHT [OUTER] JOIN support

### ISSUE THREAD

**Issue by panarch** [gluesql/gluesql#63]
For now, GlueSQL supports 2 join types, `INNER JOIN` and `LEFT OUTER JOIN`.
`RIGHT OUTER JOIN` Implementation can be a bit more tough, but all modifications required to change is in a single file, 198 lines of codes.
* [src/executor/join.rs](https://github.com/gluesql/gluesql/blob/main/src/executor/join.rs)

It can be a sort of quite fun programming quiz.

**Comment by willy610** [comment]
I have an implementation on join (all kinds (outer,inner,left and right)) in both ``Mathematica`` and ``Javascript``

They could work on ordered input or not.
I would like to try to implement that in Rust.

I hope someone is not at the moment working in ``join.rs``

In any case the solution today is slow

**Comment by panarch** [comment]
Sounds good, then I'll assign you on this issue.

**Comment by willy610** [comment]
Some feedback 

Have a working joiner now.
Left, right, inner, outer etc
(Sofar 242 lines!!)

Slightly prepared for using future existing ordered indexes...

Some notes/question and/or suggestions

1. In the ``select.rs`` perhaps one should introduce a ``'from'`` function
replacing all join stuff in ``select.rs`` and perform all tasks in the ``'from' (TableWithJoins)``
And of cource ``join.rs`` becomes something else (partly moved to inside ``'from'``)

2. I think all selections producing 'relations' should have sufficient content
so that a temporary relation could be constructed and asked for 
Like in ``execute.rs Statement::CreateTable + Statement::Insert``
(So there will be an entry in fetch like ``fetch_columns_defs()`` )

3. The projection in ``select.rs picks named and columns and must find them in the result from 'from'

See alos issue #83

Made a reverse of AST for Select into an BNF and than to a railed bnf in attachment


[sql_selectsvg.pdf](https://github.com/gluesql/gluesql/files/5342404/sql_selectsvg.pdf)

**Comment by panarch** [comment]
I think I don't get it, can you provide more details about `from` function suggestion?
It sounds like similar with current structs named `BlendContext`, `FilterContext` and `AggregateContext`.

**Comment by willy610** [comment]
I will look more into the design pattern Context.
It is complicated I think but powerful.

I'll be back

**Comment by willy610** [comment]
Well, I think I have come out with a working solution a from component.
The new ``from.fetchblended()`` will do the complete join of 0..n tables
so the old ``join`` will be replaced.
But
When removing the ``join`` step I got problem in the ``aggregate`` step
Please anyone help me.

In the ``executor/select.rs`` the old fragment looks like
```
    let rows = fetch_blended(storage, table, columns)?
        .flat_map(move |blend_context| {
            let join_columns = Rc::clone(&join_columns);
            join.apply(blend_context, join_columns)
        })
        .filter_map(move |blend_context| {
            blend_context.map_or_else(
                |error| Some(Err(error)),
                |blend_context| {
                    filter
                        .check_blended(&blend_context)
                        .map(|pass| pass.as_some(blend_context))
                        .transpose()
                },
            )
        })
        .enumerate()
        .filter_map(move |(i, item)| limit.check(i).as_some(item));

    let rows = {
        let rows = aggregate.apply(rows)?;
        let rows = Box::new(rows);

        rows.map(move |blend_context| blend.apply(blend_context))
    };

    Ok(rows)
```



In that part as above I remove the ``join`` thing.
The ``fetch_blended`` and ``from.fetchblended`` have the same return signature
and will both cause compiler error on ``aggregate.apply(rows)``
NEW VERSION with problem
```
//    let rows = fetch_blended(storage, table, columns)?
	let rows = from.fetchblended()?

//        .flat_map(move |blend_context| {
//            let join_columns = Rc::clone(&join_columns);
//            join.apply(blend_context, join_columns)
//        })
        .filter_map(move |blend_context| {
            blend_context.map_or_else(
                |error| Some(Err(error)),
                |blend_context| {
                    filter
                        .check_blended(&blend_context)
                        .map(|pass| pass.as_some(blend_context))
                        .transpose()
                },
            )
        })
        .enumerate()
        .filter_map(move |(i, item)| limit.check(i).as_some(item));

    let rows = {
        let rows = aggregate.apply(rows)?;
        let rows = Box::new(rows);

        rows.map(move |blend_context| blend.apply(blend_context))
    };

    Ok(rows)
```
Compiler says
```
301 |                 let rows = aggregate.apply(rows)?;
    |                                      ^^^^^ expected struct `executor::context::blend_context::BlendContext`, found struct `std::rc::Rc`
    |
    = note: expected enum `std::option::Option<std::result::Result<executor::context::blend_context::BlendContext<'_>, _>>`
               found enum `std::option::Option<std::result::Result<std::rc::Rc<executor::context::blend_context::BlendContext<'_>>, _>>`

```

**Comment by panarch** [comment]
It is little hard to follow, compile error looks simple which is about type conflict one is `Rc<BlendContext<_>>`, the other is `BlendContext<_>`
And whatever changes you make, `cargo test` will verify it's acceptable or not.

It would be good to call `cargo test` frequently during you make code changes.
Now I'm not sure but you look you start by changing to many codes.

**Comment by willy610** [comment]
1. Well I will wind up the question myself as time goes by, I think.
2.1 I really rely on ``cargo test`` . Very safely
2.2 I do have my own ``cargo test --bin integrationtestN`` too
3. Yes. "...start by changing to many codes changes". 
I think I manage to schedule the suggested changes.. 

I did some mistake in the GitHub desktop yes but
I work of course on copies of the my fork. 
Focus is now on ``from`` with all ``JOIN``. 
Must resolve column names in that very source, but will not propagate that in the first commit.

Question:
I this thread a forum for outlines of the design of the ``from``


By the way I been working with software development for 45 years. Dev, prod, support. In big companies.
There has been a lot of different code management processes during that time...
I will behave more properly!!! and take some more lessons

**Comment by panarch** [comment]
I don't worry any of your skills at all. Though there can be some communication issues, I think that's not a big deal.
While I was keeping eye on your changes, I found `Context` can be too unnecessarily complicated.
I thought it might really bother you to understand internal code logic.

I did some work https://github.com/gluesql/gluesql/pull/100
I simply removed `UnionContext` and `Context` in `blend.rs`.

And, soon I think it's also possible to merge two different contexts `FilterContext` and `BlendContext` into a single `Context`.

Hope you this can be helpful to you.

**Comment by willy610** [comment]
Just for the record.
Working like an idiot.
Challenge. No row copy, clone etc. 
Complex work but fun

**Comment by panarch** [comment]
> Complex work but fun
I completely agree.

**Comment by willy610** [comment]
I have been working on a copy of ``gluesql-main-3``

Some points.
1. One source called ``executor/from.rs`` on like 650 loc
2. Used in ``executor/selct.rs`` through
`` let _the_from = From::new(storage, &table_with_joins).compile().execute();``
3. Struggling with letting ``_the_from`` giving an proper iterator via ``_the_form.fetch_blended()``
4. Executes all statements in unit test besides
  4.1 ``ON`` must of kind ``{left,oper,right}`` where left and right is a table and col ref.
 Oper is ``EQ`` or ``AND``
  4.2 Only ``tbl.col`` accepted as left or right. Perhaps a permanent restriction.
   4.3 So ``INNER JOIN Player p9 ON p9.id = Item.player_id AND Item.id > 101`` will not work. Move it to the where clause.
  4.4 So `` JOIN Arith b ON a.id = b.id + 1`` will not work. Can be rewritten
  4.5 So ``JOIN BlendItem i ON u.id = 1 AND u.id = i.player_id`` fails to. Can be rewritten
  4.6 ``USING`` not supported (yet). Or doubtful?
  4.7 ``NATURAL JOIN`` not supported. Never? Ambiguous!
5. All detected errors will ``panic!()`` for the time being
6. Brief outline of implementation
  6.1 Compile the Abstract Syntax Tree AST into a executable form. Here we can rearrange the order of join depending on future optimal planner.
  6.2 Execute the joins via plain ``fold``. At each step build an index, or use an existing future index,  with key from one or more columns and two bags with the key: one bag for left rows and one for right rows. Scan left leg, from previous fold step, and put records in the left bag. Scan the right, new, table and put rows in the right leg. Then scan the index. Make a Cartesion product of left and right bag for one key. Depending on join type (Left, Inner etc) concat left and rigth row or null.
  6.3 Each step will produce rows where each row is a vector with one original ``Rc:row`` from the storage.
  6.4 Finally each row as in 6.3 is flattend into a result ``Row``

And

How to continue? 
I need feedback. Is it suitable? Does it fits in? Some parts need refactoring and styling.
Must use Vector instead of Vec. Lack debug and format in there

I can produce zip on my project. Or what is most suitably?
It's to early to work on some fork I think.

Ok
/willy

**Comment by panarch** [comment]
> 4.3 So INNER JOIN Player p9 ON p9.id = Item.player_id AND Item.id > 101 will not work. Move it to the where clause.
This kind of queries are quite common, some complex join conditions can be replaced by `WHERE` but some cannot.

> 4.6 USING not supported (yet). Or doubtful?
Currently GlueSQL does not support primary key feature, after that it will become possible, but not for now.

>4.7 NATURAL JOIN not supported. Never? Ambiguous!
You can work on natural join by making a different issue, this is for implementing right join. So this should only include right join related codes.

>6.1 Compile the Abstract Syntax Tree AST into a executable form. Here we can rearrange the order of join depending on future optimal planner.
If there exists special reason that we cannot support RIGHT JOIN without implementing executable form, then it makes sense.
However, it is possible to implement RIGHT JOIN without executable form, then the feature should not be included in this issue.

It is important to find minimal changes required to implement RIGHT JOIN.
During working on this, I guess you got lots of inspiration and motivation to do further works.
If that changes are very necessary to work on this issue because it blocks RIGHT JOIN implementation, then you can do it at this point.

So, after you implement minimal version which supports RIGHT JOIN, then feel free to make PR and I'll review that.

**Comment by willy610** [comment]
Thanks for your comments.
Will look into RIGHT JOIN.

**Comment by willy610** [comment]
Perhaps this could be a solution in a nusthell

Use but refine the fold inside ``excutor/join.rs`` to work with a local ``fn proj_prod(outer_rows,inner_rows``) and invoke with either
+ ``proj_prod(outer_rows,inner_rows);``

or (swapped)

+ ``proj_prod(inner_rows,outer_rows);``

and swap result if in the last call type.

Then combine the two result row bags into a new row bag to feed fold with.

Of course also pass to ``proj_prod`` whether inner or outer is requested.

The following sources are to be changed a think

1. Introduce ``excutor/from.rs`` Holds most of todays ``excutor/join.rs``
2. Move all table and join functions from ``excutor/select.rs`` into ``excutor/from.rs``
Lot of duplicate code I think.
3. Make a rough copy of join.rs into from.rs
4. Then replace the fold inside ``excutor/from.rs`` 
5. Replace the ``fn join`` inside todays ``excutor/from.rs`` with a wider ``proj_prod``

A first step could be to verify results towards SQL-statmenets and have it working in paralell with todays solution. 

And the next step could to remake it as a proper component with all filter, contetxts, etc

**Comment by willy610** [comment]
### Some notes on Join (and indexes)

1. FULL [ OUTER ] joins could be compiled into 'select ... left join(...) union right join(...)'
2. INNER join. If indexes are present chose the smallest index one for outer scan.
3. LEFT or RIGHT. Look into pseuocode here. Outer must scan whole table. ``iter()`` on inner could be a 'match' on index
4. I think moving all joins from select.rs and join.rs into a new from.rs would make things more maintainable

See also 
[Visual-Representation-of-SQL-Joins](https://www.codeproject.com/Articles/33052/Visual-Representation-of-SQL-Joins)

```
from(first_table,joins)
{
// joins is AST or some kind of optimized 'RPN'
 let final_result =
	joins.iter().fold(first_table,|joined_sofar,the_next_join_clause|{
		let Joiner{next_table,join_operator}= the_next_join_clause;
		let swap =if join_operator.contains("right"){true}else{false};
		let result_to_fold = 
		if swap {joined_sofar}else{next_table}
		.iter()
		.map(|outer_row|){
			let the_join_on_one_left_with_all_on_right = 
				if swap {next_table}else{joined_sofar}.iter()
				.filter_map(|inner_row|{
					let compare = check(outer_row + inner_row);
					// return to filter
					if compare{Some(inner_row)}else{None}
				});
			// anything joined?
			// inner join ->skip if empty
			let inner_prod = 
				if the_join_on_one_left_with_all_on_right.len() != 0
				{
					if swap
						{prod(the_join_on_one_left_with_all_on_right, outer_row}
					else
						{prod(outer_row ,the_join_on_one_left_with_all_on_right}
				}
				else
				{
					if swap
						{prod('null', outer_row}
					else
						{prod(outer_row ,'null'}
				};
		};
		flatten(result_to_fold)
	});
	final_result
}
```

**Comment by MRGRAVITY817** [comment]
Seems like my kind of stuff. I'll take this.

**Comment by MinhyukWoo** [comment]
I'd like to take this on. I spent some time reading the current `main` (a155abde) end-to-end for RIGHT JOIN support and want to share the plan before opening a PR, since this issue has been open for 5+ years and the one prior attempt (#114) got stuck on row-merging logic in a 3-table join.

## Why this isn't a small change

Adding `JoinOperator::RightOuter` touches the full pipeline (AST → translate → plan → executor → query_builder → tests → docs), and most of it is mechanical — the compiler's exhaustive `match`es walk you through it. Two spots aren't mechanical, though:

### 1. Conflicts with the primary-key point-lookup optimization (#1919, fixed again in #1943)

`PrimaryKeyPlanner::select` rewrites `WHERE pk = literal` on the **first FROM relation** into an index lookup, dropping the `WHERE` clause, regardless of join operator (see the `left_outer_join_installs_lookup_on_first_relation` test in `plan/primary_key.rs`). That's safe for INNER/LEFT because the first relation's rows are never null-padded — but for `A RIGHT JOIN B`, A *can* be null-padded, so narrowing A down early and dropping the filter changes which B-rows get treated as "unmatched." I worked through a concrete example and it produces a different row count than the correct RIGHT JOIN semantics.

`plan/primary_key/lookup.rs` already has a deliberately exhaustive match (`validate_join_operator`) with a comment ("Keep this exhaustive so new join types require an explicit lookup-safety decision") that will fail to compile once `RightOuter` is added — so this has to be decided, not silently inherited. My proposed first-PR decision: disable this optimization whenever `RightOuter` is present in the FROM chain, and leave "safely re-enable it for the preserved side" as follow-up work.

Two related spots that *won't* fail to compile — `plan/schemaless/query.rs` and `plan/schemaless/validate.rs` both have `_ => {}` catch-alls on `JoinOperatorPlan` matches, so they'd silently skip RIGHT JOIN's ON-clause rewriting/validation unless I explicitly add a `RightOuter` arm.

### 2. Executor redesign

The current `join()` in `executor/join.rs` is a left-fold over left rows; `LeftOuter` only needs to ask "did *this* left row match anything?" — a local, per-row question. RIGHT JOIN needs the opposite: "did this right/new-relation row match *any* left row, across the whole traversal?" — a global question the current structure has no place to track (NestedLoop rescans per left row with no shared state; Hash only does per-row key lookups).

Proposed design: materialize the new relation once (`Vec<Row>`), track matched rows by **position index** in a shared `HashSet` (not by join-key, to avoid falsely marking a whole hash bucket as matched when some rows fail the `WHERE` clause), build a null-context template per join clause in parallel with real rows, then `chain()` an "unmatched" pass after the main traversal that emits `(null_template, row)` for every index never marked seen. This applies uniformly to NestedLoop and Hash — Hash just needs the bucket lookup to also record into the same `seen` set — so I don't think there's a good reason to punt Hash support to a follow-up.

(Swapping `A RIGHT JOIN B` → `B LEFT JOIN A` at translate time was tempting but doesn't work here: it just relocates the PK-optimization bug to B, and `plan/join.rs`'s `table_with_joins` accumulates context strictly left-to-right, so it doesn't compose when a RIGHT JOIN is in the middle of a longer join chain.)

## Proposed scope for the first PR

- 2-table `A RIGHT JOIN B ON ...`, both NestedLoop and Hash condition shapes
- PK point-lookup optimization disabled whenever `RightOuter` is present (safe default, not a full fix)
- Explicit `RightOuter` handling in the two schemaless catch-alls above
- `query_builder` `right_join`/`right_join_as` builders
- Fixture tests (equality + non-equality conditions, empty-left-table edge case, RIGHT JOIN + GROUP BY) and planner unit tests for the PK/Hash decisions
- Docs update

Follow-ups I'd explicitly defer: RIGHT JOIN in the middle of 3+ table chains (should work by construction, but I'd rather verify separately given #1919→#1943 already showed a "should work in theory" PK optimization silently breaking on multi-relation chains), and safely re-enabling the PK optimization for RIGHT JOIN's preserved side.

Let me know if this approach (or the scope split) looks off before I start — happy to adjust.

**Comment by panarch** [comment]
Thanks for the detailed analysis. The PK lookup, schemaless handling, and matched-right-row tracking concerns are all valid. Before starting the implementation, however, I would like to adjust the direction based on the execution-plan work introduced in #1937.

The general direction is to move decisions that can be made during planning out of the executor. Schemaless and aggregate handling have already moved in this direction: planners produce explicit execution-plan states, while the executor only manages runtime state.

## Join planning states

The current join planner is effectively a hash join planner: `NestedLoop` is an executable default, and the planner replaces it with `Hash` when it finds a valid hash candidate. Renaming it accordingly is tracked separately in #1966.

I think this default should remain unchanged for INNER and LEFT OUTER joins:

```text
NestedLoop
→ HashJoinPlanner
    ├─ hash candidate found: Hash
    └─ no candidate: keep NestedLoop
```

Hash planning is an optional optimization. Failure to find a candidate is not an error because `NestedLoop` remains executable.

RIGHT OUTER JOIN is different. Before its semantics have been planned, it is not yet an executable join with the current plan structure. I suggest representing only this state explicitly, for example as `UnplannedRightOuter`, rather than introducing a global `Unplanned` executor state for every join.

```text
AST RightOuter
→ UnplannedRightOuter
→ RightOuterJoinPlanner
    ├─ successfully lowered: executable RightOuter plan
    └─ cannot be lowered: unsupported/planner error
```

Validation and schemaless planning can still inspect and rewrite the relation and `ON` expression while it is in this state. The RightOuter planner would be a no-op when no RIGHT JOIN exists, but lowering is required when `UnplannedRightOuter` is present. No unplanned RIGHT JOIN should reach later planners or the executor.

## Planner order

The intended order would be:

```text
validate
→ schemaless
→ right outer join
→ primary key
→ index
→ hash join
→ aggregate
```

The RightOuter planner needs to run before the primary-key planner because a RIGHT JOIN requires the complete accumulated left input to determine whether preserved right rows are matched.

For example, pushing `WHERE A.pk = 1` into the scan of `A` in `A RIGHT JOIN B` can remove A rows that should have matched B rows. Those B rows can then be emitted incorrectly as unmatched. Instead of making `PrimaryKeyPlanner` reinterpret raw RIGHT JOIN syntax, the planned RightOuter state should tell later planners that the complete left input is required. The safe initial behavior is to skip the root PK lookup in that case.

After this semantic lowering, `NestedLoop` can remain the default execution strategy. The HashJoin planner may later replace it with `Hash`, just as it does for existing joins.

## RightOuter planner and executor responsibilities

The RightOuter planner should determine:

- that the new/right relation is the preserved side;
- that matching requires the complete accumulated left input;
- which accumulated relations must be NULL-extended for an unmatched right row;
- the expressions required for matching and later hash optimization;
- the PK lookup restriction required by the join.

The executor should only manage runtime state:

```text
prepare right rows
→ probe accumulated left rows
→ record matched right-row positions
→ emit unmatched right rows
→ done
```

Tracking by row position is important because duplicate rows and residual `ON` predicates can cause rows in the same hash bucket to have different match results. For a hash execution plan, all right rows must also remain available for the unmatched pass, including rows with NULL hash keys or rows excluded from the hash candidate index by a right-side predicate.

A multi-join case should be included in the initial regression coverage. In:

```sql
A
JOIN B ON ...
RIGHT JOIN C ON ...
```

an unmatched C row requires the accumulated A and B prefix to be NULL-extended. This is also the area where the previous attempt in #114 ran into row-composition problems, so deferring all 3-table verification would leave the most important structural case untested.

## Query builder

I would treat the query builder API as a separate design decision. The builder currently constructs `StatementPlan` directly and also exposes physical-plan methods such as `hash_executor` and `hash_filter`. Because of that, adding `right_join` and `right_join_as` is not only a syntax-level API addition; we also need to decide how an unplanned RightOuter state interacts with explicit hash plans.

The SQL planner/executor support can proceed first, while the RightOuter query builder API and its interaction with explicit execution-strategy methods are handled as follow-up work.

**Comment by MinhyukWoo** [comment]
Thanks for the detailed redirect — this lines up with the direction #1937 already took for schemaless/aggregate. Before starting, one direction question and a scope note:

## 1. State representation

Rather than a separate `UnplannedRightOuter` variant that becomes a distinct `RightOuter` variant after lowering, I'm planning a single variant with an `Option` field:
```rust
pub enum JoinOperatorPlan {
    Inner(JoinConstraintPlan),
    LeftOuter(JoinConstraintPlan),
    RightOuter {
        constraint: JoinConstraintPlan,
        null_extend: Option<Vec<NullExtendRelation>>,  // None until lowered
    },
}
```
`schemaless`, `primary_key`, `index`, the hash-candidate planner, and `aggregate` all match `RightOuter { constraint, .. }` and never touch `null_extend`. Only the executor reads it, returning `Err` (not panic — GlueSQL embeds via FFI/WASM) if it's still `None` when reached, which should only happen if the pipeline wiring itself is broken.

I did look at splitting this into two distinct types instead — an `UnplannedRightOuter` variant that pre-lowering plan stages carry, separate from an executor-facing `RightOuter` variant that would only exist after lowering succeeds, so the executor's `match` couldn't even compile against an unlowered state — but `StatementPlan`/`QueryPlan`/`ExprPlan` recursively reference each other through subqueries, so that split touches on the order of 15-30 files crate-wide for a guarantee a linear pipeline order already gives in practice. I'd rather defer it unless you feel strongly otherwise.

## 2. `plan/index.rs`

`plan/index.rs` has the same root-relation-narrowing problem as the PK planner, just via secondary indexes: its WHERE-clause rewrite doesn't look at `join_operator` at all, so `WHERE A.indexed_col = x` gets turned into an index seek and the predicate gets dropped even when `A` is RIGHT JOIN's null-padded side — a silent wrong row count, not an error, and there's no existing test that would catch it. I'll gate that rewrite behind the same lookup-safety check as the PK fix (skip it whenever `RightOuter` is in `from.joins`), leaving the separate ORDER BY-based index optimization untouched since it only reorders the scan and never drops rows.

## Proposed scope for this PR

- `right_outer_join` planner (new), including recursion into FROM-clause derived-table subqueries so a RIGHT JOIN nested inside one is lowered the same as at the top level, + wiring into both planner pipelines between `schemaless` and `primary_key`
- PK/index root-relation lookup disabled whenever `RightOuter` is present in the FROM chain
- New `RightOuter` executable variant in `executor/join.rs`, matched by row position so it applies uniformly to NestedLoop and Hash
- Regression coverage: 3+ table chain with RIGHT JOIN mid-chain (where #114 got stuck), RIGHT JOIN nested inside a derived-table subquery, Hash/NestedLoop condition shapes, PK/index lookup regressions, RIGHT JOIN + GROUP BY, RIGHT JOIN ON-subquery referencing a schemaless table, correlated subquery on RIGHT JOIN's right side
- `validate()`: relation alias column-count and subquery wildcard-alias checks
- Docs update

Deferred:
- `right_join`/`right_join_as` query builder and its interaction with explicit hash-plan methods, as you noted
- Safely re-enabling PK/index lookup for RIGHT JOIN's preserved side

Separately: `plan/join.rs` (hash-candidate planner) doesn't recurse into FROM-clause derived-table subqueries at all today, for any join type — unrelated to RIGHT JOIN and not a correctness issue (NestedLoop stays valid, it just misses the Hash optimization), so I'll file it as its own issue rather than fold it into this PR.

Heads-up: #1937 (draft, unmerged) touches several of the same files — flagging for rebase awareness, not blocking.

Does this look right to start?

**Comment by panarch** [comment]
Thanks for thinking this through so carefully. Reading your proposed representation, I think the difference between our models is what we expect to survive the AST-to-plan boundary.

I believe you are treating `RightOuter` similarly to the existing `Inner` and `LeftOuter` variants: the operator remains in `JoinOperatorPlan` through execution, while `null_extend: None` or `Some(...)` indicates whether right-outer planning has completed. That is an understandable reading of the current structure, and my phrase “executable RightOuter plan” may also have made it sound as though an executor-facing `RightOuter` variant should remain.

What I intended was slightly different:

```text
AST RightOuter
→ UnplannedRightOuter
→ validate
→ schemaless
→ RightOuterJoinPlanner
→ fully lowered join plan
→ primary key
→ index
→ hash join
→ aggregate
→ executor
```

`RightOuter` remains an AST concept. Translation converts it into the temporary logical-plan state `UnplannedRightOuter`. Validation and schemaless planning can inspect and rewrite that state, and then `RightOuterJoinPlanner` consumes it completely.

The later planners and executor receive only the fully lowered join plan, in which the preserved-side and NULL-extension decisions are already represented as required plan data. They do not receive either `UnplannedRightOuter` or an executor-facing `RightOuter` variant.

With that model, this representation is not needed:

```rust
RightOuter {
    constraint: JoinConstraintPlan,
    null_extend: Option<Vec<NullExtendRelation>>,
}
```

It would keep the AST-level `RightOuter` concept throughout the pipeline and use `None`/`Some` to distinguish two different planning phases. I would instead like the right-outer planner to perform an actual lowering and remove the temporary state. `null_extend`, or whatever planned structure represents the same decision, should be mandatory after that boundary rather than optional.

I am also not proposing two separate recursive pre-planning and executor-facing plan type families in this PR.

Some additional architectural context may help explain why I am drawing the boundary there. The plan structures we have recently introduced after the AST currently function primarily as a logical plan and shared planning IR. Ideally, the architecture would eventually have another boundary:

```text
AST
→ Logical Plan
→ Concrete Execution Plan
→ Executor
```

In that design, the concrete execution plan would be valid by construction. It would contain no `Unplanned` variants or optional markers for incomplete planning phases, and the executor could only receive fully executable states.

We have only recently introduced the logical-plan structure and have not yet built that additional concrete execution-plan layer. For now, the same plan representation must carry a small number of intermediate planning states. Allowing this one explicit `UnplannedRightOuter` variant to exist temporarily between translation and the right-outer planner is the practical compromise in the current architecture.

The important part is that the temporary state has a clear lifetime and a clear consumer. When a separate concrete execution-plan structure is introduced later, intermediate states such as this can move entirely out of the executor-facing representation. There is still plenty of worthwhile plan-architecture work ahead of us, which is a good situation to be in, but this PR does not need to introduce that entire additional layer.

Regarding #1937, please do not worry about the overlapping files or let that constrain or block this work. I can coordinate the overlap and any necessary rebase or integration afterward. Please focus on making the RIGHT OUTER JOIN change internally complete and correct.

I also would not use the number of changed files as the scope criterion. A coherent change may touch one line or a hundred files. What matters is whether the PR has a clear beginning and end and forms one complete unit of work. If validation or planning changes are prerequisites for making that unit correct, they belong in the PR. If they are independent cleanup or optimization work, they should remain separate.

With that clarification, the proposed scope and overall direction look good to start.
